llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
          SubReg.getReg() == SuperVec.getReg();
 }
 
+// Verify the illgal copy from VGPR to SGPR for generic opcode COPY
+bool SIInstrInfo::verifyCopy(const MachineInstr &MI,
+                             const MachineRegisterInfo &MRI,
+                             StringRef &ErrInfo) const {
+  const MachineOperand &Dst = MI.getOperand(0);
+  const MachineOperand &Src = MI.getOperand(1);
+
+  if (Dst.isReg() && Src.isReg()) {
+    Register DstReg = Dst.getReg();
+    Register SrcReg = Src.getReg();
+    // This is a check for copy from an VGPR to SGPR
+    if (RI.isVGPR(MRI, SrcReg) && RI.isSGPRReg(MRI, DstReg)) {
+      ErrInfo = "illegal copy from VGPR to SGPR";
+      return false;
+    }
+  }
+  return true;
+}
+
 bool SIInstrInfo::verifyInstruction(const MachineInstr &MI,
                                     StringRef &ErrInfo) const {
   uint16_t Opcode = MI.getOpcode();
-  if (SIInstrInfo::isGenericOpcode(MI.getOpcode()))
-    return true;
-
   const MachineFunction *MF = MI.getParent()->getParent();
   const MachineRegisterInfo &MRI = MF->getRegInfo();
 
+  if (SIInstrInfo::isGenericOpcode(MI.getOpcode())) {
+    // FIXME: At this point the COPY verify is done only for non-ssa forms.
+    // Find a better property to recognize the point where instruction selection
+    // is just done.
+    if (!MRI.isSSA() && MI.isCopy())
```

Comment:
Why is the isSSA check needed?

Reply:
Because we lack the appropriate property to express the real condition, and this is a half measure. We really want a custom IsSelectedAndSGPRCopiesFixed property, otherwise the natural check will fail before SIFixSGPRCopies

Reply:
OK, then the comment should say explicitly say that we can only enforce this after SIFixSGPRCopies.

Reply:
I was wondering; the existing ` MachineFunctionProperties::Property::Selected` currently set in the GISel pipeline when selection is completed, should be set in the SelectionDAG as well (mostly after finalize-isel) for consistency. MachineFunctions currently don't have this property set for SelectionDAG.
If this field is rightly set, we could catch the illegal copies soon after finalize-isel, by then SIFixSGPRCopies would have been invoked.

Reply:
The problem is if selectiondag sets it, the natural place it would be set in selection is still before SIFixSGPRCopies

Reply:
Isn't FinalizeISel that marks the end of selection? If not, the pass name is misleading.

Reply:
Yes and no, and I don't think it's really misleading. It finalizes the already selected function into a verifiable form

Reply:
Improve the comment to include Jay's suggestion.
"explicitly say that we can only enforce this after SIFixSGPRCopies."

Reply:
Improved the comment.